solver/src/interactions/allowances.rs

```diff
@@ -0,0 +1,291 @@
+
+        Ok(if allowance < amount {
+            Approval::Approve {
+                token,
+                spender: self.spender,
+            }
+        } else {
+            Approval::Sufficient
+        })
+    }
+}
+
+/// An ERC20 approval interaction.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Approval {
+    /// The exisisting allowance is sufficient, so no additional `approve` is
+    /// required.
+    Sufficient,
+
+    /// An ERC20 approve is needed. This interaction always approves U256::MAX
+    /// in order to save gas by allowing approvals to be used over multiple
+    /// settlements.
+    Approve { token: H160, spender: H160 },
+}
+
+impl Interaction for Approval {
+    fn encode(&self) -> Vec<EncodedInteraction> {
+        match self {
+            Approval::Sufficient => vec![],
+            Approval::Approve { token, spender } => {
+                // Use a "dummy" contract - unfortunately `ethcontract` doesn't
+                // allow you use the generated contract intances to encode
+                // transaction data without a `Web3` instance. Hopefully, this
+                // limitation will be lifted soon to clean up stuff like this.
+                let token = dummy_contract!(ERC20, *token);
+                let approve = Erc20ApproveInteraction {
+                    token,
+                    spender: *spender,
+                    amount: U256::max_value(),
+                };
+
+                approve.encode()
+            }
+        }
+    }
+}
+
+/// An allowance manager that an address.
+pub struct AllowanceManager {
+    web3: Web3,
+    owner: H160,
+}
+
+#[async_trait::async_trait]
+impl AllowanceManaging for AllowanceManager {
+    async fn get_allowances(&self, tokens: HashSet<H160>, spender: H160) -> Allowances {
+        Allowances::new(
+            spender,
+            fetch_allowances(self.web3.clone(), tokens, self.owner, spender).await,
+        )
+    }
+}
+
+async fn fetch_allowances<T>(
+    web3: ethcontract::Web3<T>,
+    tokens: HashSet<H160>,
+    owner: H160,
+    spender: H160,
+) -> HashMap<H160, U256>
+where
+    T: ethcontract::web3::BatchTransport + Send + Sync + 'static,
+    T::Out: Send,
+{
+    let mut batch = CallBatch::new(web3.transport());
+    let results: Vec<_> = tokens
+        .into_iter()
+        .map(|token| {
+            let allowance = ERC20::at(&web3, token)
+                .allowance(owner, spender)
+                .batch_call(&mut batch);
+            (token, allowance)
+        })
+        .collect();
+
+    let _ = batch.execute_all(MAX_BATCH_SIZE).await;
+
+    let mut allowances = HashMap::new();
+    for (token, allowance) in results {
+        let allowance = match allowance.await {
+            Ok(value) => value,
+            Err(err) => {
+                tracing::warn!("error retrieving allowance for token {:?}: {}", token, err);
+                continue;
+            }
```

Comment:
Is `continue` the right reaction to an error? This means we treat failure to get allowance the same as no allowance. If all of the request failed because of a transport error we would return an empty HashMap instead of an error.
Maybe it makes sense to return Result::Err on any non smart contract error and only continue on smart contract errrors.

Reply:
The main reason I did this was that this is how allowances were treated in the current Uniswap liquidity implementation:
https://github.com/gnosis/gp-v2-services/blob/0bca0d5a4e05fcd8ace2ea4cf5e97ad5d555285b/solver/src/liquidity/uniswap.rs#L108-L112

Additionally, this method would return an `Allowances` "cache" instance that treats missing allowances as errors, and different then tokens with allowance 0:

https://github.com/gnosis/gp-v2-services/blob/c29d89a506d768d5603e2dd146aa432fdc1347a9/solver/src/interactions/allowances.rs#L53

So I thought this was a happy medium where we could cache allowances ignoring errors (so it doesn't error out for tokens that end up not getting used at all) and then potentially produce errors when trying to make an `Approval` interaction for a token whose allowance we were unable to retrieve.

Reply:
I think a single token failure shouldn't make the entire request fail. Maybe we can return an error if the top level batch request fails?

Reply:
Added slightly hacky support to detect batch errors. I [created an issue in `ethcontract`](https://github.com/gnosis/ethcontract-rs) to improve batch error handling in order to make the detection more reliable.

Reply:
My suggestion for a fix would have been to replace

```rust
batch.execute_all(MAX_BATCH_SIZE).await;
```

with
```rust
batch.execute_all(MAX_BATCH_SIZE).await?;
```

If the batch itself fails the top level execute_all would fail wouldn't it?

Reply:
[`execute_all` doesn't return a result](https://docs.rs/ethcontract/0.12.2/ethcontract/batch/struct.CallBatch.html#method.execute_all) and seems to just forward errors to each call future.